### PR TITLE
Fix unit tests: tfw_str

### PR DESCRIPTION
**Pull request by ikoveshnikov**
fix few issues in unit tests for tfw_str.

### CODE REVIEW

File: tempesta_fw/t/unit/test_tfw_str.c [hunk 1]
```diff
@@ -416,6 +416,10 @@ TEST(tfw_str_eq_cstr_pos, plain)
 	TfwStr *fox = make_plain_str(foxstr), *c, *end;
 	long i, offset = 0, foxlen = fox->len;
 
+	/* unlike to tfw_str_eq_cstr_off we don't  know which pointer we should use
+	 * to address string element at 'offset' position from begging of the string.
+	 * go though chunks, manually to get needed pointers
+	 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_tfw_str.c]: The comment is completely unnecessary. The behaviour is implied by the nature of the tested function itself.
In this case, however, the test is against a plain string, so it doesn't make sense to go over chunks of the string as there are none.

File: tempesta_fw/t/unit/test_tfw_str.c [hunk 2]
```diff
@@ -432,31 +436,35 @@ TEST(tfw_str_eq_cstr_pos, plain)
 		}
 	}
 
+	/* Addressing substring which is actually not a part of string will
+	 * raise an error. That is expected behaviour and we should not worried
+	 * about that
+	*/
+	TEST_LOG("Ignore next warning: \"Desired position is outside the string\". "
+	         "We are checking invalid argument value\n");
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_tfw_str.c]: Frankly, this looks ugly. Perhaps in this particular case it's best to remove the warning from `tfw_str_eq_cstr_pos()` and return -1? Then whoever is interested in that case can see if that's what happened.

File: tempesta_fw/t/unit/test_tfw_str.c [hunk 3]
```diff
@@ -471,6 +479,10 @@ TEST(tfw_str_eq_cstr_pos, compound)
 	TfwStr *fox = make_compound_str(foxstr), *c, *end;
 	long i, offset = 0, foxlen = fox->len;
 
+	/* unlike to tfw_str_eq_cstr_off we don't  know which pointer we should use
+	 * to address string element at 'offset' position from begging of the string.
+	 * go though chunks, manually to get needed pointers
+	 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_tfw_str.c]: I don't believe the comment is needed.

File: tempesta_fw/t/unit/test_tfw_str.c [hunk 4]
```diff
@@ -487,30 +499,35 @@ TEST(tfw_str_eq_cstr_pos, compound)
 		}
 	}
 
+	/* Addressing substring which is actually not a part of string will
+	 * raise an error. That is expected behaviour and we should not worried
+	 * about that
+	*/
+	TEST_LOG("Ignore next warning: \"Desired position is outside the string\". "
+	         "We are checking invalid argument value\n");
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_tfw_str.c]: Again, this does look ugly...

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 5]
```diff
@@ -42,102 +42,131 @@
 #include "server.h"
 #include "test.h"
 
-TEST(tfw_sched_rr, sg_empty)
+static TfwMsg *
+sched_rr_get_scheduler_arg (size_t connection_type __attribute__((unused)))
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: Space between the function name and the opening brace.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 6]
```diff
@@ -42,102 +42,131 @@
...
 {
-	int i;
+	return NULL;
+}
 
-	TfwSrvGroup *sg = test_create_sg("test", "round-robin");
+static void
+sched_rr_free_scheduler_arg (TfwMsg * msg __attribute__((unused)))
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: Space between the function name and the opening brace.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 7]
```diff
@@ -42,102 +42,131 @@
...
-	TfwSrvGroup *sg = test_create_sg("test", "round-robin");
+	TfwSrvGroup *sg = test_create_sg("test", sched_helper_rr.scheduler);
 	TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 	for (i = 0; i < TFW_SRV_MAX_CONN; ++i) {
 		TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
 		sg->sched->add_conn(sg, srv, &sconn->conn);
-		s ^= (long long)sconn;
+		conn_acc ^= (long long)&sconn->conn;
+		/* connection reference count must be >0
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: The first line of a multi-line comment should contain just a single `/*`. Then please start the sentence with a capital character.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 8]
```diff
@@ -42,102 +42,131 @@
...
+		 * otherwise connection will be scipped in scheduler
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: Perhaps, `scipped` is `skipped`? Using a spell checker might be a good idea.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 9]
```diff
@@ -43,96 +43,92 @@
...
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
+	TfwSrvGroup *sg = test_create_sg("test", sched_helper_hash.scheduler);
 	TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 	for (i = 0; i < TFW_SRV_MAX_CONN; ++i) {
 		TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
 		sg->sched->add_conn(sg, srv, &sconn->conn);
+		/* connection reference count must be >0
+		 * otherwise connection will be scipped in scheduler
+		 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: The first line of a multi-line comment should contain just a single /*. Then please start the sentence with a capital character. This applies to all comments within functions in this patch set.
`scipped` -> `skipped`.

File: tempesta_fw/t/unit/sched_helper.c [hunk 10]
```diff
@@ -133,3 +134,84 @@ test_conn_release_all(TfwSrvGroup *sg)
 		}
 	}
 }
+
+void test_sched_generic_empty_sg(struct TestSchedHelper * sched_helper)
```
Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: Please don't put a space between the asterisk `*` and the name of a variable. This applies to many places in this patch set.

Also, please put the function name on a separate line. This also applies to multiple places.

File: tempesta_fw/t/unit/sched_helper.c [hunk 11]
```diff
@@ -133,3 +134,84 @@ test_conn_release_all(TfwSrvGroup *sg)
...
+	BUG_ON(!sched_helper->scheduler);
+	BUG_ON(!sched_helper->connections_types);
+	BUG_ON(!sched_helper->get_scheduler_arg);
+	BUG_ON(!sched_helper->free_scheduler_arg);
+
+	sg = test_create_sg("test", sched_helper->scheduler);
+
+	for (i = 0; i < sched_helper->connections_types; ++i) {
+		for (j = 0; j < 3; ++j) {
+			TfwMsg * msg = sched_helper->get_scheduler_arg(i);
```
Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: Space between the asterisk and the name of a variable.

File: tempesta_fw/t/unit/sched_helper.c [hunk 12]
```diff
@@ -133,3 +134,84 @@ test_conn_release_all(TfwSrvGroup *sg)
...
+	BUG_ON(!sched_helper->free_scheduler_arg);
+
+	sg = test_create_sg("test", sched_helper->scheduler);
+
+	for (j = 0; j < TFW_SG_MAX_SRV; ++j) {
+		test_create_srv("127.0.0.1", sg);
+	}
+
+	for (i = 0; i < sched_helper->connections_types; ++i) {
+		for (j = 0; j < 2 * TFW_SG_MAX_SRV; ++j) {
```
Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: What is the point of trying twice as many times that there's servers in the group? Same question applies to the above two functions, where the test is tried three times.
There's similar code in other parts of this patch set.

Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: What is the point of trying twice as many times that there're servers in the group? Same question applies to the above two functions, where the test is tried three times.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 13]
```diff
@@ -43,96 +43,92 @@
...
+static char *req_strs[] = {
+	"GET / HTTP/1.1\r\nhost:host1\r\n\r\n",
+	"GET / HTTP/1.1\r\nhost:host2\r\n\r\n",
+	"GET / HTTP/1.1\r\nhost:host3\r\n\r\n",
+	"GET / HTTP/1.1\r\nhost:host4\r\n\r\n",
 };
-size_t req_strs_size = sizeof(req_strs) / sizeof(req_strs[0]);
 
-TEST(tfw_sched_hash, sg_empty)
+static TfwMsg * sched_hash_get_scheduler_arg (size_t connection_type);
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: Space between the function name and the opening brace. This applies to multiple places in this patch set.
Also, please put the function name on a separate line. This also applies to multiple places.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 14]
```diff
@@ -43,96 +43,92 @@
...
+sched_hash_free_scheduler_arg (TfwMsg * msg)
 {
-	int i, j;
+	test_req_free((TfwHttpReq *)msg);
+}
 
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
+static struct TestSchedHelper sched_helper_hash = {
+	.scheduler = "hash",
+	.connections_types = sizeof(req_strs) / sizeof(req_strs[0]),
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: Perhaps, `ARRAY_SIZE(res_strs)`?

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 15]
```diff
@@ -43,96 +43,92 @@
...
+	.get_scheduler_arg = &sched_hash_get_scheduler_arg,
+	.free_scheduler_arg = &sched_hash_free_scheduler_arg,
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: The function names and the variable names look way too long. In this particular case, I would suggest something like this:
`.scheduler` -> `.sched`
`.connection_types` -> `.conn_types`
`.get_scheduler_arg` -> `.get_sched_arg`
`sched_hash_get_scheduler_arg` -> `sched_hash_get_arg`.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 16]
```diff
@@ -43,96 +43,92 @@
...
-
-			tfw_http_parse_req(req, req_strs[i], strlen(req_strs[i]));
+			TfwMsg *msg = sched_helper_hash.get_scheduler_arg(i);
+			TfwConnection *conn = sg->sched->sched_srv(msg, sg);
+			EXPECT_NOT_NULL(conn);
 
-			conn = sg->sched->sched_srv((TfwMsg *)req, sg);
-			if (!s) {
-				s = conn;
+			if (!expected_connection) {
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: No brackets are needed if a conditional block consists of a single statement.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 17]
```diff
@@ -42,102 +42,131 @@
...
 		TfwServer *srv = test_create_srv("127.0.0.1", sg);
 		for (j = 0; j < TFW_SRV_MAX_CONN; ++j) {
-			sconn = test_create_conn((TfwPeer *)srv);
+			TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
 			sg->sched->add_conn(sg, srv, &sconn->conn);
-			s ^= (long long)sconn;
+			conn_acc ^= (long long)&(sconn->conn);
+			/* connection reference count must be >0
+			 * otherwise connection will be scipped in scheduler
+			 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: `/* A connection is skipped by schedulers if (refcnt <= 0). */`

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 18]
```diff
@@ -142,64 +138,51 @@ TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
...
+			/* connection reference count must be >0
+			 * otherwise connection will be scipped in scheduler
+			 */
+			tfw_connection_get(&sconn->conn);
 		}
 	}
 
-	for (i = 0; i < req_strs_size; ++i) {
-		TfwConnection *s = NULL;
+	/*	Check if every single request is scheduled to the same connection
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: `/* Check that every request is scheduled to the same connection. */

File: tempesta_fw/cfg.c [hunk 19]
```diff
@@ -1315,7 +1322,8 @@ tfw_cfg_set_str(TfwCfgSpec *cs, TfwCfgEntry *e)
 
 		min = cse->len_range.min;
 		max = cse->len_range.max;
-		if (min != max && (len < min || len > max)) {
+		if (((min != 0) && (max != 0)) &&
+			(len < min || len > max)) {
```
Comment (ikoveshnikov) [inline tempesta_fw/cfg.c]: Yes, you're right, i must have missed that comment. I looked though tempesta configuration once again and exact string length is not one of our use case. I will remove that code modifications and related unit test.

File: tempesta_fw/t/unit/test_tfw_str.c (hunk 1 above)
Comment (ikoveshnikov) [inline tempesta_fw/t/unit/test_tfw_str.c]: We raised question about complexity of this unit test in conference twice last week. May be it should  be commented?
About going over chunks of a plain string. Do we really need to make unit test  rely on implementation? I should have think about that before, but compound  and plain strings have the same interface, so unit tests are mostly the same for both. May be it is better to provide a single test function and pass compound or plain strings as parameter?

File: tempesta_fw/t/unit/sched_helper.h [hunk 20]
```diff
@@ -40,4 +40,16 @@ TfwSrvConnection *test_create_conn(TfwPeer *peer);
...
+	size_t conn_types;
+	TfwMsg * (*get_sched_arg)(size_t connection_type);
+	void (*free_sched_arg)(TfwMsg *);
+};
+
+void test_sched_generic_empty_sg(struct TestSchedHelper *sched_helper);
+void test_sched_generic_one_srv_in_sg_zero_conn(struct TestSchedHelper *sched_helper);
+void test_sched_generic_max_srv_in_sg_and_zero_conn(struct TestSchedHelper *sched_helper);
+
+
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/sched_helper.h]: extra empty line

File: tempesta_fw/t/unit/sched_helper.c [hunk 21]
```diff
@@ -133,3 +134,82 @@ test_conn_release_all(TfwSrvGroup *sg)
...
+		TfwConnection *conn = sg->sched->sched_srv(msg, sg);
+
+		EXPECT_NULL(conn);
+		sched_helper->free_sched_arg(msg);
+	}
+
+	test_sg_release_all();
+}
+
+void test_sched_generic_max_srv_in_sg_and_zero_conn(struct TestSchedHelper *sched_helper)
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/sched_helper.c]: Too long line, please don't exceed 80 characters per line. In fact the function name is just enormous large. Also we write function type and its name on different lines. Please read https://github.com/tempesta-tech/linux-4.1.27-tfw/blob/master/Documentation/CodingStyle

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 22]
```diff
@@ -142,64 +136,50 @@ TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
...
 
 	for (i = 0; i < TFW_SG_MAX_SRV; ++i) {
 		TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 		for (j = 0; j < TFW_SRV_MAX_CONN; ++j) {
-			sconn = test_create_conn((TfwPeer *)srv);
+			TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
+
 			sg->sched->add_conn(sg, srv, &sconn->conn);
+			/* A connection is skipped by schedulers if (refcnt <= 0). */
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/test_sched_hash.c]: Too long line

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 23]
```diff
@@ -142,64 +136,50 @@ TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
...
+			tfw_connection_get(&sconn->conn);
 		}
 	}
 
-	for (i = 0; i < req_strs_size; ++i) {
-		TfwConnection *s = NULL;
+	/* Check that every request is scheduled to the same connection. */
+	for (i = 0; i < sched_helper_hash.conn_types; ++i) {
+		TfwConnection *expected_connection = NULL;
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/test_sched_hash.c]: Not a problem, but usually we don't use long names for functions and variables, for example `exp_conn`'d be better.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 24]
```diff
@@ -142,64 +136,50 @@ TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
...
 	for (i = 0; i < TFW_SG_MAX_SRV; ++i) {
 		TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 		for (j = 0; j < TFW_SRV_MAX_CONN; ++j) {
-			sconn = test_create_conn((TfwPeer *)srv);
+			TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
+
 			sg->sched->add_conn(sg, srv, &sconn->conn);
+			/* A connection is skipped by schedulers if (refcnt <= 0). */
+			tfw_connection_get(&sconn->conn);
```
Comment (krizhanovsky) [inline tempesta_fw/t/unit/test_sched_hash.c]: Shouldn't we just set connection reference counter to 1 in `test_creaate_conn()`?

File: tempesta_fw/t/unit/sched_helper.c [hunk 25]
```diff
@@ -133,3 +134,82 @@ test_conn_release_all(TfwSrvGroup *sg)
 		}
 	}
 }
+
+void test_sched_generic_empty_sg(struct TestSchedHelper *sched_helper)
```
Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: Please place the function type and the function name on separate lines.

File: tempesta_fw/t/unit/sched_helper.c [hunk 26]
```diff
@@ -133,3 +134,82 @@ test_conn_release_all(TfwSrvGroup *sg)
...
+		TfwConnection *conn = sg->sched->sched_srv(msg, sg);
+
+		EXPECT_NULL(conn);
+		sched_helper->free_sched_arg(msg);
+	}
+
+	test_sg_release_all();
+}
+
+void test_sched_generic_one_srv_in_sg_zero_conn(struct TestSchedHelper *sched_helper)
```
Comment (keshonok) [inline tempesta_fw/t/unit/sched_helper.c]: Please place the function type and the function name on separate lines.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 27]
```diff
@@ -43,96 +43,90 @@
...
-	int i, j;
+	test_req_free((TfwHttpReq *)msg);
+}
 
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
+static struct TestSchedHelper sched_helper_hash = {
+	.sched = "hash",
+	.conn_types = ARRAY_SIZE(req_strs),
+	.get_sched_arg = &sched_hash_get_scheduler_arg,
+	.free_sched_arg = &sched_hash_free_scheduler_arg,
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: Isn't this some sort of tauthology? `sched_hash` is the common part of the name, so we know that it's related to a scheduler, and moreover, to the hash scheduler. So when we want to get/free an arg, what's the point in saying that it's a `scheduler_arg`? Just `sched_had_get/free_arg` would be sufficient, and make the name much shorter and concise.

File: tempesta_fw/t/unit/test_sched_hash.c [hunk 28]
```diff
@@ -43,96 +43,90 @@
...
 
-			tfw_http_parse_req(req, req_strs[i],
-					   strlen(req_strs[i]));
+	BUG_ON(connection_type >= sched_helper_hash.conn_types);
 
-			conn = sg->sched->sched_srv((TfwMsg *)req, sg);
-			EXPECT_TRUE(conn == NULL);
+	req = test_req_alloc(strlen(req_strs[connection_type]));
+	tfw_http_parse_req(req, (unsigned char *) req_strs[connection_type],
+					   strlen(req_strs[connection_type]));
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_hash.c]: Not a problem in this particular case, but more often than not long names are a burden as code statements don't fit a single line. `conn_type` is simpler and shorter without sacrificing the meaning.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 29]
```diff
@@ -42,102 +42,131 @@
...
 	}
 
-	for (i = 0; i < 3 * TFW_SRV_MAX_CONN; ++i) {
-		TfwConnection *conn = sg->sched->sched_srv(NULL, sg);
-		s ^= (long long)conn;
-		tfw_connection_put(conn);
-	}
+	/* Check that connections is scheduled in the fair way:
+	 * every connection will be scheduled only once
+	 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: Please fix the style of a multi-line comment within a function. That had been commented in the review.

File: tempesta_fw/t/unit/test_sched_rr.c [hunk 30]
```diff
@@ -42,102 +42,131 @@
...
 	}
 
-	for (j = 0; j < 3 * TFW_SG_MAX_SRV * TFW_SRV_MAX_CONN; ++j) {
-		TfwConnection *conn = sg->sched->sched_srv(NULL, sg);
-		s ^= (long long)conn;
-		tfw_connection_put(conn);
-	}
+	/* Check that connections is scheduled in the fair way:
+	 * every connection will be scheduled only once
+	 */
```
Comment (keshonok) [inline tempesta_fw/t/unit/test_sched_rr.c]: Wrong multi-line comment style.

**Pull request merged**